### ISSUE TITLE
`toxlsx` fails when header is not a string

### ISSUE THREAD

**Issue by scharf** [petl-developers/petl#530]
#### Minimal, reproducible code sample, a copy-pastable example if possible

The [documentation says](https://petl.readthedocs.io/en/stable/intro.html?highlight=header%20str#conventions-table-containers-and-table-iterators):

> a *header value* is typically a string (`str`) but may be an object of any type as long as it implements `__str__` and is pickleable

Saving `xlsx` fails when the header is no string:

```python

def test_toxlsx_with_non_str_header():
    class Header:
        def __init__(self, name):
            self.__name = name

        def __str__(self):
            return self.__name

    tbl = ((Header('foo'), Header('bar')),
           ('A', 1),
           ('B', 2),
           ('C', 2))
    f = NamedTemporaryFile(delete=True, suffix='.xlsx')
    f.close()

    toxlsx(tbl, f.name, 'Sheet1')
```

#### Version and installation information

* Value of ``petl.__version__`` `1.6.8`
* Version of Python interpreter `3.8.1`
* Operating system : Mac
* How petl was installed (e.g., "using pip into virtual environment", or "using conda"): `virtual environment`

**Comment by juarezr** [comment]
@scharf,

There is no handling or conversion on `petl` side because it just fowards the "rows" to `openxyl`:

``` python
    for row in rows:
        ws.append(row)
```

Also on `openxyl` I haven't found handling for that:

``` python
        for col_idx, value in enumerate(values, 1):
            if value is None:
                continue
            try:
                cell.value = value
            except ValueError:
                if isinstance(value, Cell):
                    cell = value
                else:
                    raise ValueError
```

A transformation before `toxlsx()` would solve this for you?

``` python
def _cast_header_as_str(tbl):
    for row_idx, row in enumerate(tbl, 1):
        if row_idx == 1:
            yield([str(col) from col in row])
        yield(row)

tbl2 = _cast_header_as_str(tbl)

toxlsx(tbl, fname, 'Sheet1')

```

**Comment by scharf** [comment]
> There is no handling or conversion on petl side because it just fowards the "rows" to openxyl:

Then it would be good to add this inconsistency to the documentation. Other exports do the conversion

- `todb` [does the conversion](https://github.com/petl-developers/petl/blob/aa18cb52c471650b4149b680a6ab82fb7a0b4abe/petl/io/db.py#L411-L413)
- `totext` [does the conversion](https://github.com/petl-developers/petl/blob/ee769a7229083231278e66e5020d8a80424729b8/petl/io/text.py#L196-L198)
-  `tobcolz` [does the conversion](https://github.com/petl-developers/petl/blob/65139a519165555ccf9b621b9a08a66e60c4a462/petl/io/bcolz.py#L139-L141)
-  `toavro ` [does the conversion](https://github.com/petl-developers/petl/blob/d713928c79fe8cd7406835ccf324ca557dc0c732/petl/io/avro.py#L562-L564)
- `tojson`, `tojsonarrays`, `tohtml`, `todataframe` do the conversion implicitly
- `topickle` *does not do the conversion to string* but that is expected. It stores the table as is and when loading it recovers the objects (this is the second condition of headers because they are pickable)...

It seems that `toxlsx` is the only exception.

> Also on openxyl I haven't found handling for that

This is not the point. The point is that `petl` is a layer on top of those libraries and it promises that everything works and it is the job of `petl` to do the conversion and not of the underlying libraries.

Therefore, I think my proposed test is a valid expectation after reading the documentation and the code...

I have solved it this way (which is about 10 time faster (though this does not really matter, because IO is slow)):

```python
from petl.compat import next, text_type
import itertools

def header_as_str(table):
    """
    Converts the header to text

    :param table: a petl table
    :return: an iterator with the first row as text
    """
    it = iter(table)
    hdr = next(it)
    flds = list(map(text_type, hdr))
    return itertools.chain((flds,), it)
```

**Comment by juarezr** [comment]
@scharf 

Good points.

I'm not able to comment about if this would be the `job` of `petl` in this case. Maybe @alimanfoo could help here.

Maybe this case and some others rough edges would benefit from a patch. I don't know also if it's possible to handle most cases in a acceptable way.

**Comment by arturponinski** [comment]
Merged into master, this will be released in [v1.7.8](https://github.com/petl-developers/petl/milestone/33)
